components/allocator/tracking.rs

```diff
@@ -0,0 +1,211 @@
+    /// The start of the allocated memory.
+    ptr: *mut u8,
+    /// The size of the allocated memory.
+    size: usize,
+    /// True if this allocation site's size is ever queried after the initial
+    /// allocation. If false, it means that the allocation is not visible from
+    /// any of the MallocSizeOf roots.
+    noted: bool,
+}
+
+impl AllocSite {
+    fn contains(&self, ptr: *mut u8) -> bool {
+        ptr >= self.ptr && ptr < unsafe { self.ptr.offset(self.size as isize) }
+    }
+}
+
+unsafe impl Send for AllocSite {}
+
+/// A map of pointers to allocation callsite metadata.
+static ALLOCATION_SITES: Mutex<FnvHashMap<usize, AllocSite>> =
+    const { Mutex::new(FnvHashMap::with_hasher(FnvBuildHasher::new())) };
+
+#[derive(Default)]
+pub struct AccountingAlloc<A = System> {
+    allocator: A,
+}
+
+impl<A> AccountingAlloc<A> {
+    pub const fn with_allocator(allocator: A) -> Self {
+        Self { allocator }
+    }
+
+    fn remove_allocation(&self, ptr: *const c_void, size: usize) {
+        if size < MIN_SIZE {
+            return;
+        }
+        let old = IN_ALLOCATION.with(|status| status.replace(true));
+        if old {
+            return;
+        }
+        let mut sites = ALLOCATION_SITES.lock().unwrap();
+        if let Entry::Occupied(e) = sites.entry(ptr as usize) {
+            e.remove();
+        }
+        IN_ALLOCATION.with(|status| status.set(old));
+    }
+
+    fn record_allocation(&self, ptr: *mut u8, size: usize) {
+        if size < MIN_SIZE {
+            return;
+        }
+        let old = IN_ALLOCATION.with(|status| status.replace(true));
+        if old {
+            return;
+        }
+        let mut num_skipped = 0;
+        let mut num_frames = 0;
+        let mut frames = [std::ptr::null_mut(); MAX_FRAMES];
+        backtrace::trace(|frame| {
+            if num_skipped < SKIPPED_FRAMES {
+                num_skipped += 1;
+                return true;
+            }
+            frames[num_frames] = frame.ip();
+            num_frames += 1;
+            num_frames < MAX_FRAMES
+        });
+        let site = AllocSite {
+            frames,
+            size,
+            ptr,
+            noted: false,
+        };
+        let mut sites = ALLOCATION_SITES.lock().unwrap();
+
+        if sites.len() < MAX_TRACKED_ALLOCATIONS {
+            sites.insert(ptr as usize, site);
+        } else if let Some(key) = sites
+            .iter()
+            .find(|(_, s)| s.size < site.size)
+            .map(|(k, _)| k.clone())
+        {
+            sites.remove(&key);
+            sites.insert(ptr as usize, site);
+        }
+
+        IN_ALLOCATION.with(|status| status.set(old));
+    }
+
+    pub(crate) fn enclosing_size(&self, ptr: *const c_void) -> (*const c_void, usize) {
+        ALLOCATION_SITES
+            .lock()
+            .unwrap()
+            .iter()
+            .find(|(_, site)| site.contains(ptr.cast_mut().cast()))
+            .map_or((std::ptr::null_mut(), 0), |(_, site)| {
```

Comment:
Is the `or` case only possible when `ptr` is used after free, or are there also other possiblities? All ptrs we receive here, should have been allocated from the global rust allocator, right (excluding UB)?

Reply:
I wonder if we might want to make this a debug assertion then. Perhaps logging an error would be preferable, but I'm not sure if allocations when logging might be problematic or not.

Reply:
I can do this as a followup.